src/layout.rs

```diff
@@ -366,140 +366,144 @@ impl Layout {
+                    element.size() | EQ(WEAK) | f64::from(m),
+                ])?;
+            }
+            Constraint::Min(m) => {
+                solver.add_constraints(&[
+                    element.size() | GE(STRONG) | f64::from(m),
+                    element.size() | EQ(WEAK) | f64::from(m),
+                ])?;
             }
         }
     }
-    solver.add_constraints(&ccs).unwrap();
+
     let changes: HashMap<Variable, f64> = solver.fetch_changes().iter().copied().collect();
+
+    // convert to Rects
     let mut results = elements
         .iter()
-        .map(|element| Rect {
-            x: changes.get(&element.x).map(|&v| v as u16).unwrap_or(0),
-            y: changes.get(&element.y).map(|&v| v as u16).unwrap_or(0),
-            width: changes.get(&element.width).map(|&v| v as u16).unwrap_or(0),
-            height: changes.get(&element.height).map(|&v| v as u16).unwrap_or(0),
+        .map(|element| {
+            let x = *changes.get(&element.x).unwrap_or(&0.0);
+            let y = *changes.get(&element.y).unwrap_or(&0.0);
+            let width = *changes.get(&element.width).unwrap_or(&0.0);
+            let height = *changes.get(&element.height).unwrap_or(&0.0);
+            Rect {
+                x: x as u16,
+                y: y as u16,
+                width: width as u16,
+                height: height as u16,
+            }
         })
         .collect::<Rc<[Rect]>>();
 
     if layout.expand_to_fill {
-        // Fix imprecision by extending the last item a bit if necessary
-        // "unwrap" is safe, because the Rc at this point has no shared references
+        // Because it's not always possible to satisfy all constraints, the last item might be
+        // slightly smaller than the available space. E.g. when the available space is 10, and the
+        // constraints are [Length(5), Max(4)], the last item will be 4 wide instead of 5. Fix this
+        // by extending the last item a bit if necessary. (`unwrap()` is safe here, because results
+        // has no shared references right now).
         if let Some(last) = Rc::get_mut(&mut results).unwrap().last_mut() {
             match layout.direction {
-                Direction::Vertical => {
-                    last.height = dest_area.bottom() - last.y;
-                }
                 Direction::Horizontal => {
-                    last.width = dest_area.right() - last.x;
+                    last.width = inner.right() - last.x;
+                }
+                Direction::Vertical => {
+                    last.height = inner.bottom() - last.y;
                 }
             }
         }
     }
 
-    results
+    Ok(results)
 }
 
 /// A container used by the solver inside split
 #[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
-struct Element {
+pub(crate) struct Element {
+    direction: Direction,
     x: Variable,
     y: Variable,
     width: Variable,
     height: Variable,
 }
```

Comment:
I like this simplification, and I'm perfectly fine with it this way. But I _think_ you can go one step further.

For `Horizontal` layouts the resulting `Rect`s `top` and `bottom` are always equal to the `top` and `bottom` of the target area. Similarly, for `Vertical`, `left` and `right` are always equal to the target area's `left` and `right`.

In the code you have at the moment, you are running it through the algorithm:

```rust
    for element in &elements {
        solver.add_constraints(&[
            element.cross_start() | EQ(REQUIRED) | area.cross_start(),
            element.cross_end() | EQ(REQUIRED) | area.cross_end(),
        ])?;
    }
```

But you can probably just set the values on the rect in the result directly, based on direction.

If you do that, I _think_ `Element` can be simplified further to just:

```rust
struct Element {
  coordinate: Variable,
  length: Variable,
}
```

Reply:
Neat - I'll give that a try